nix/servant-prometheus.nix

```diff
@@ -106,7 +106,7 @@ in { system, compiler, flags, pkgs, hsPkgs, pkgconfPkgs, ... }:
     } // {
     src = (pkgs.lib).mkDefault (pkgs.fetchgit {
       url = "https://github.com/PrivateStorageio/servant-prometheus.git";
-      rev = "ec21c5ed50e6f6f8e52916ce71cd68fcd0166cad";
+      rev = "b1207d687aeb48bd035f986e88287391e4937e23";
       sha256 = "0lswszfs52x5rpf7lj46iv77zghcbr4d05dwssi63yzjll1ixizd";
```

Comment:
I changed rev but not the hash, hoping I would get an error report with the new hash.
Instead it just worked.
What's wrong?

Reply:
This one bit me so many times.  It's a really awful UX.  What happened is that Nix looked at your hash, looked in the store for a matching object, found it, and used it.  You told it that whatever you wanted, it would have that hash, so whatever it finds with a matching hash must be what you wanted.

There are two workflows that I generally follow, either of which will avoid this:

* When you change rev (or anything else that means you want a different source), corrupt the hash by replacing one of the digits with something else (it still has to be a syntactically valid hash or nix will complain, though).  Then, when it complains that what it fetched does not match the hash you gave it, it will tell you the hash of the thing it got.  Copy that into the source and it'll work on the next build.
* Use one of the `nix-prefetch-...` tools to get the source you want.  It will tell you the hash of what it got.  Copy that into the source and it'll build what you want next time.
*

Reply:
> You told it that whatever you wanted, it would have that hash, so whatever it finds with a matching hash must be what you wanted.

😹 Those programmers are such jokesters. Thanks for hinting me at it!

Reply:
It does the same and seems to be working with the current hash, a corrupt hash and with removing that line completely.

Could it be that it doesn't use that file but gets the revision spec'd in [stack.yaml](https://github.com/PrivateStorageio/PaymentServer/pull/90/files#diff-0de01f189d384bfbe69e806ef70b1a425b27394dbe2a9d0e62ffd010dda86c03) instead?

Reply:
"Researching the world wide web" [I learned about nix-prefetch-git](https://github.com/NixOS/nix/issues/1880#issuecomment-367187213)

Reply:
Ah crap.  Which build system are you using? :/

Reply:
Ahem.  To elaborate ...

If you use `stack build` then whatever is in `stack.yaml` will be respected (and all of `nix/` ignored).  If you use `nix-build nix/ -A ...` then all of the things in `nix/` will be respected (and `stack.yaml` ignored).

`stack build` does a bit less redundant work so it's more convenient for the normal development workflow.  But `nix-build ...` needs to work so that everything integrates properly in the end.  Everything in `nix/` is auto-generated by some version of `stack-to-nix` from haskell.nix (which has been changing faster than I can keep up with, so very likely the *newest* stack-to-nix will not work without some other fixes).

Reply:
Why wouldn't we have two build systems 😹

OK, that explains things! Thank you!